### PR TITLE
Add support for Kalray VLIW archictecture aka KVX

### PR DESCRIPTION
**Pull request by jmaselbas**
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional) [test-bin#71](https://github.com/radareorg/radare2-testbins/pull/71)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

Add support for Kalray VLIW archictecture aka KVX

KVX is VLIW architecture, kind of like a DSP, it can execute more than
one "instruction" per cycle. Each cycle a bundle of instruction is
executed which is comprised of up to 6 instructions. The general rules
for bundle are: only one BCU instruction (which comes first); only one
LSU instruction; and up to 4 ALU instructions (depending on the type).

In assembly the end of a bundle is marked by a `;;` in an empty line, here
in r2 (to make things more compact and readable) the end of bundle is
appended after the last instruction, but beware that this is not valid
kvx assembly.

### CODE REVIEW

**Comment by jmaselbas** [comment]
I have have some questions on the build system, I don't know how to properly test the meson build (even less for the windows builds) and I might I've missed changes from "merging" the asm to anal pluging

File: libr/anal/p/anal_kvx.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	if (op == NULL) {
+		return 0;
+	}
+
+	if (addr % 4) {
+		goto unaligned;
+	}
+
+	insn = kvx_next_insn (&bundle, addr, b, len);
+	if (insn == NULL) {
```
Comment (trufae) [inline libr/anal/p/anal_kvx.c:237]:
```suggestion
	if (!insn) {
```

File: libr/anal/p/anal_kvx.c [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+	if (op == NULL) {
+		return 0;
+	}
+
+	if (addr % 4) {
+		goto unaligned;
+	}
+
+	insn = kvx_next_insn (&bundle, addr, b, len);
```
Comment (trufae) [inline libr/anal/p/anal_kvx.c:236]:
```suggestion
	insn_t *insn = kvx_next_insn (&bundle, addr, b, len);
```

File: libr/anal/p/anal_kvx.c [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+"gpr	ea	.64 1600 0\n"
+"gpr	ev	.64 1632 0\n"
+"gpr	sr	.64 1664 0\n"
+"gpr	es	.64 1696 0\n"
+	;
+
+static int kvx_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *b, int len, RAnalOpMask mask) {
+	char strasm[64];
+	static bundle_t bundle = { };
+	insn_t *insn;
```
Comment (trufae) [inline libr/anal/p/anal_kvx.c:226]:
```suggestion
```

**Review by trufae — CHANGES_REQUESTED** [review]
Looks good to me! is there anything that doesnt works for you?

**Comment by trufae** [comment]
../libr/anal/p/anal_kvx.c(227): error C2059: syntax error: '}'

File: libr/anal/arch/kvx/kvx-dis.c [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+static int read_bundle(ut32 *words, const ut8 *buf, int len) {
+	int count = 0;
+	ut32 word;
+
+	while (len >= sizeof (ut32) && count < KVX_MAX_BUNDLE_WORDS) {
+		word = r_read_le32 (buf);
+		words[count] = word;
+		count++;
+
+		if (!kvx_has_parallel_bit (word))
```
Comment (trufae) [inline libr/anal/arch/kvx/kvx-dis.c:218]: missing braces {}

File: libr/anal/arch/kvx/kvx-dis.c [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	return count;
+}
+
+insn_t *kvx_next_insn(bundle_t *bundle, ut64 addr, const ut8 *buf, int len) {
+	ut32 words[KVX_MAX_BUNDLE_WORDS];
+	ut64 start;
+	int issue, count;
+	int ret;
+
+	issue = KVX_MAX_BUNDLE_ISSUE;
```
Comment (trufae) [inline libr/anal/arch/kvx/kvx-dis.c:238]:
```suggestion
	int issue = KVX_MAX_BUNDLE_ISSUE;
```
define and assign in the same line

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by jmaselbas** [comment]
```../libr/anal/p/anal_kvx.c(227): error C2059: syntax error: '}'```
That might be a compiler bug, looks like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53119 
the struct is initialized with `= { };` even tho this isn't needed because it's declared as static; I will remove the `= { };`

File: libr/anal/p/anal_kvx.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+"gpr	sps	.64 1568 0\n"
+"gpr	ea	.64 1600 0\n"
+"gpr	ev	.64 1632 0\n"
+"gpr	sr	.64 1664 0\n"
+"gpr	es	.64 1696 0\n"
+	;
+
+static int kvx_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *b, int len, RAnalOpMask mask) {
+	char strasm[64];
+	static bundle_t bundle = { };
```
[the remark this answers is not in the document]
Comment (jmaselbas) [reply libr/anal/p/anal_kvx.c]: yes it must be static... that's a hack. this in order to return each instruction of a bundle individually...
That's ugly but I've not found a better way of doing this.
OK will make the change to `static R_TH_LOCAL`

File: libr/anal/arch/kvx/kvx-dis.c [hunk 7]
```diff
@@ -0,0 +1,275 @@
+/* radare - LGPL - Copyright 2022 - jmaselbas */
```
[the remark this answers is not in the document]
Comment (jmaselbas) [reply libr/anal/arch/kvx/kvx-dis.c:1]: I didnt get the question.

[the remark this answers is not in the document]
Comment (trufae) [reply libr/anal/arch/kvx/kvx-dis.c:1]: the plugin description specifies "GPL" as the license for this plugin, so the header should reflect that, if its coming from binutils it shold be GPL, but if its a clean/rework reimplementaion for r2 you can pick any other license. just trying to clarify that before the debian maintainers complain :D

**Comment by trufae** [comment]
Looks good now!

File: libr/anal/arch/kvx/kvx-dis.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jmaselbas) [reply libr/anal/arch/kvx/kvx-dis.c:1]: it's a reimplementation for r2. Only the `disassemble_bundle` function is a rework of an existing version from binutils.

**Comment by jmaselbas** [comment]
OK I happy with this version being merged

**Pull request merged**